arangod/Aql/IResearchViewOptimizerRules.cpp

```diff
@@ -120,6 +122,117 @@ bool optimizeSearchCondition(IResearchViewNode& viewNode, Query& query, Executio
 
   return true;
 }
+    
+bool optimizeSort(IResearchViewNode& viewNode, ExecutionPlan* plan) {
+  std::unordered_map<VariableId, AstNode const*> variableDefinitions;
+
+  ExecutionNode* current = static_cast<ExecutionNode*>(&viewNode);
+
+  while (true) {
+    current = current->getFirstParent();
+
+    if (current == nullptr) {
+      // we are at the bottom end of the plan
+      return false;
+    }
+      
+    if (current->getType() == EN::ENUMERATE_IRESEARCH_VIEW ||
+        current->getType() == EN::ENUMERATE_COLLECTION ||
+        current->getType() == EN::TRAVERSAL ||
+        current->getType() == EN::SHORTEST_PATH ||
+        current->getType() == EN::INDEX ||
+        current->getType() == EN::COLLECT) {
+      // any of these node types will lead to more/less results in the output,
+      // and may as well change the sort order, so let's better abort here
+      return false;
+    }
+        
+    if (current->getType() == EN::CALCULATION) {
+      // pick up the meanings of variables as we walk the plan
+      variableDefinitions.emplace(
+          ExecutionNode::castTo<CalculationNode const*>(current)->outVariable()->id,
+          ExecutionNode::castTo<CalculationNode const*>(current)->expression()->node());
+    }
+
+    if (current->getType() != EN::SORT) {
+      // from here on, we are only interested in sorts
+      continue;
+    }
+    
+    std::vector<std::pair<Variable const*, bool>> sorts;
+
+    auto* sortNode = ExecutionNode::castTo<SortNode*>(current);
+    auto const& sortElements = sortNode->elements();
+
+    sorts.reserve(sortElements.size());
+    for (auto& it : sortElements) {
+      // note: in contrast to regular indexes, views support sorting in different 
+      // directions for multiple fields (e.g. SORT doc.a ASC, doc.b DESC). 
+      // this is not supported by indexes
+      sorts.emplace_back(it.var, it.ascending);
+    }
+
+    SortCondition sortCondition(plan, 
+                                sorts,
+                                std::vector<std::vector<arangodb::basics::AttributeName>>(),
+                                variableDefinitions);
+
+    if (sortCondition.isEmpty() || !sortCondition.isOnlyAttributeAccess()) {
+      // unusable sort condition
+      return false;
+    }
+
+    auto& view = arangodb::LogicalView::cast<IResearchView>(*viewNode.view());
+    auto& primarySort = view.primarySort();
+    
+    // sort condition found, and sorting only by attributes!
+
+    if (sortCondition.numAttributes() > primarySort.size()) {
+      // the SORT condition in the query has more attributes than the view
+      // is sorted by. we cannot optimize in this case
+      return false;
+    }
+    
+    // check if all sort conditions match
+    for (size_t i = 0; i < sortElements.size(); ++i) {
+      if (sortElements[i].ascending != primarySort.directon(i)) {
```

Comment:
should be `direction`
```suggestion
      if (sortElements[i].ascending != primarySort.direction(i)) {
```